### PR TITLE
Adds railings and fences

### PR DESCRIPTION
**Pull request by S34NW**
## What Does This PR Do
This PR adds two types of constructible objects, both made from metal rods.

The railing: A cheap, waist-high structure that blocks movement like directional glass, but can always be climbed over provided the other side is clear. Sprite source as far as I know is from `MMMiracles` over on TG, I made some minor adjustments.

The chainlink fence: An expensive, tall, un-climbable structure that blocks movement. You can cut holes in it with a wirecutter. A small hole can be slowly climbed through, a large hole can be freely walked through. There is also a door structure that can easily permit access. Sprites by `unid15` over at /VG/
Chainlink fences can be shocked like grilles with a wire node.
You can prevent users cutting through them with wire cutters via var editing the `invulnerable` var.

## Why It's Good For The Game
More variety in barricade-style structures is good, these can be used for movement control, decoration, as a barrier from dangerous terrain, or for defence.

### Chainlink fences

Open door, intact
2021-09-10 15_45_59-Paradise Station 13

Closed door, holes cut
2021-09-10 15_47_03-Paradise Station 13

### Railings
2021-09-10 16_31_27-Paradise Station 13

### Build menu
2021-09-10 15_47_31-Paradise Station 13
2021-09-10 15_47_37-Paradise Station 13

## Changelog
:cl:
add: Added chainlink fences
add: Added railings
/:cl:

### CODE REVIEW

**Comment by TDSSS** [comment]
What do holes look like on the fence when it's facing east or west?

**Comment by Wojz** [comment]
Love this. I can already see the potential to make cool stuff using it.

**Comment by S34NW** [comment]
> 
> 
> What do holes look like on the fence when it's facing east or west?

image

File: code/game/atoms.dm [hunk 1]
```diff
@@ -1153,3 +1153,10 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
 	if(degrees)
 		appearance_flags |= PIXEL_SCALE
 	transform = M
+
+///Setter for the `density` variable to append behavior related to its changing.
+/atom/proc/set_density(new_value)
+	if(density == new_value)
+		return
+	. = density
+	density = new_value
```
Comment (FlattestGuitar) [inline code/game/atoms.dm:1162]: what *is* this proc?

it returns the previous density only if it changed density, otherwise it returns false

Comment (S34NW) [reply code/game/atoms.dm:1162]: If the density is already the same value as you are trying to set it, it does nothing, but if the new value is different to the current value it sets it as that.

**Comment by Darkmight9** [comment]
Just need to be able to build basket ball hoops and you'd have a full on basket ball court :D !

File: code/game/objects/structures/railings.dm [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	if(I.use_tool(src, user, volume = 75, extra_checks = CALLBACK(src, .proc/check_anchored, anchored)))
+		if(!anchored)
+			anchored = TRUE
+		else
+			anchored = FALSE
+		to_chat(user, "<span class='notice'>You [anchored ? "fasten the railing to":"unfasten the railing from"] the floor.</span>")
+	return TRUE
+
+/obj/structure/railing/CanPass(atom/movable/mover, turf/target)
+	if(istype(mover) && mover.checkpass(PASSTABLE))
```
Comment (S34NW) [inline code/game/objects/structures/railings.dm:68]:
```suggestion
	if(istype(mover) && mover.checkpass(PASSFENCE))
```

File: code/game/atoms.dm [hunk 3]
```diff
@@ -1153,3 +1153,10 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
 	if(degrees)
 		appearance_flags |= PIXEL_SCALE
 	transform = M
+
+///Setter for the `density` variable to append behavior related to its changing.
+/atom/proc/set_density(new_value)
+	if(density == new_value)
```
Comment (Aurorablade) [inline code/game/atoms.dm:1159]: Spess

File: code/game/objects/structures/fence.dm [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+/obj/structure/fence/cut/medium
+	icon_state = "straight_cut2"
+	hole_size = MEDIUM_HOLE
+
+/obj/structure/fence/cut/large
+	icon_state = "straight_cut3"
+	hole_size = LARGE_HOLE
+
+/obj/structure/fence/CanPass(atom/movable/mover, turf/target)
+	if(istype(mover) && mover.checkpass(PASSFENCE))
```
Comment (Aurorablade) [inline code/game/objects/structures/fence.dm:60]: Spess...

File: code/game/objects/structures/fence.dm [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		return TRUE
+	if(istype(mover, /obj/item/projectile))
```
Comment (Aurorablade) [inline code/game/objects/structures/fence.dm:62]: 👀

File: code/game/objects/structures/fence.dm [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+		return TRUE
+	if(!density)
```
Comment (Aurorablade) [inline code/game/objects/structures/fence.dm:64]: s p a c e (look i am trying to make this fun)

File: code/game/objects/structures/fence.dm [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+		return TRUE
+	return FALSE
+
+// shock user with probability prb (if all connections & power are working)
+// returns 1 if shocked, 0 otherwise
+// Totally not stolen from code\game\objects\structures\grille.dm
+
+/obj/structure/fence/proc/shock(mob/user, prb)
+	if(!prob(prb))
```
Comment (Aurorablade) [inline code/game/objects/structures/fence.dm:73]: ^

File: code/game/objects/structures/fence.dm [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+		return FALSE
+	if(!in_range(src, user)) //To prevent TK and mech users from getting shocked
```
Comment (Aurorablade) [inline code/game/objects/structures/fence.dm:75]: Spees

File: code/game/objects/structures/railings.dm [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+		to_chat(user, "<span class='warning'>[src] cannot be rotated in that direction!</span>")
+		return FALSE
+	return TRUE
+
+/obj/structure/railing/proc/check_anchored(checked_anchored)
+	if(anchored == checked_anchored)
+		return TRUE
+
+/obj/structure/railing/proc/after_rotation(mob/user,rotation_type)
+	add_fingerprint(user)
```
Comment (Aurorablade) [inline code/game/objects/structures/railings.dm:139]: I should reskim it but should we also add fingerprints for cutting or is that on the dmaging proc as well? i could see how you may not need to add it though considering it the wirecutters do it and not alien crazy knife hands.

**Comment by Aurorablade** [comment]
okay i am removing my space comments review cause apprently my brain needs coffee and REVERSED ALL THE STANDARDS

File: code/game/atoms.dm (hunk 3 above)
Comment (Aurorablade) [reply code/game/atoms.dm:1159]: i am leaving this one, but disregard it its more an eternal reminder for me to get coffee after a nap and before reviewing.

**Comment by Spacemanspark** [comment]
Interesting. I ran into these on Blue Colony. I imagine these will have a much more different use.

Would be neat to see an electrified variant of these, sort of like grilles can now.

**Comment by S34NW** [comment]
> 
> 
> Interesting. I ran into the fences on Blue Colony. I imagine these will have a much more different use, however.
> 
> Would be neat to see an electrified variant of these, sort of like grilles can now.

They can be electrified if there is a node under them 
image

**Comment by Spacemanspark** [comment]
Ah, perfect.

File: code/game/objects/structures/railings.dm [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+	if(istype(O, /obj/item/projectile))
+		return TRUE
+	if(ismob(O))
+		if(M.flying | M.floating)
+			return TRUE
+	if(O.throwing)
+		return TRUE
+	if(O.move_force >= MOVE_FORCE_EXTREMELY_STRONG)
+		return TRUE
+	if(currently_climbed)
```
[the remark this answers is not in the document]
Comment (S34NW) [reply code/game/objects/structures/railings.dm:93]: resolved with new vaulting behaviour

File: code/game/objects/structures/railings.dm (hunk 9 above)
Comment (S34NW) [reply code/game/objects/structures/railings.dm:127]: I think, as you said, the use of wirecutters negates prints

**Review by variableundefined — APPROVED** [review]

**Pull request merged**